User Submission History and User Stats

Raul9595 opened the pull request:
User Submission History and User Stats

- [x] I have added a screen http://localhost:8080/stats that will display the Submission history and the Submission Statistics of a particular user
- [x] I have added the required function def statistics() in the python file
- [x] I have added the stats.html page for rendering the page

![image](https://user-images.githubusercontent.com/16488421/58261360-c928bc80-7d95-11e9-9f8a-968795904052.png)

![image](https://user-images.githubusercontent.com/16488421/58261374-d34abb00-7d95-11e9-8bb4-05c27bb97c25.png)

rishabhKalakoti commented:
We have moved to `peewee` and `mysqli` with #126 , you should switch to database accordingly

rishabhKalakoti commented:
I was thinking to have stats for individual users (similar to profiles in online judges), I think here we have stats for all the users. @theSage21 what do you think about this? How should we progress?

Raul9595 commented:
Added User Specific Data and updated to incorporate sqlite changes
![image](https://user-images.githubusercontent.com/16488421/58278698-3a7a6680-7dba-11e9-9873-20a44768ab61.png)
![image](https://user-images.githubusercontent.com/16488421/58278724-4534fb80-7dba-11e9-9b48-a87b3320aa16.png)

theSage21 commented on server.py:
```diff
@@ -146,6 +143,38 @@ def dashboard():
+    sub_stats_total = []
+    sub_history_temp = Submission.select(Contest.code, ContestProblems.question, Submission.time, Submission.is_correct)\
+        .join(ContestProblems, on=(Submission.contestProblem == ContestProblems.id)) \
+        .join(Session, on=(Submission.user == Session.user))\
```
I think I see a problem. Try this out:

1. add 2 users
2. make some submissions for both of them
3. see stats and find your mistake

theSage21 commented on server.py:
```diff
@@ -146,6 +143,38 @@ def dashboard():
+        .switch() \
+        .join(Contest, on=(ContestProblems.contest == Contest.id)) \
+        .order_by(Submission.time.desc())
+    for sub in sub_history_temp.tuples():
```
we can simply do `sub_hist = list(sub_hist_temp.tuples())`

theSage21 commented on server.py:
```diff
@@ -146,6 +143,38 @@ def dashboard():
+        .order_by(Submission.time.desc())
+    for sub in sub_history_temp.tuples():
+        sub_history.append(sub)
+    sub_stats_correct_temp = (
```
since you already have all the submissions and their "correct/incorrect" values let's use that to construct this number instead of another database call?

Raul9595 commented:
I have updated the code as per the review comments.

theSage21 commented on server.py:
```diff
@@ -146,6 +143,20 @@ def dashboard():
+    sub_history_temp = Submission.select(Contest.code, ContestProblems.question, Submission.time, Submission.is_correct)\
+        .where(Session.token == bottle.request.get_cookie("s_id"))\
+        .join(ContestProblems, on=(Submission.contestProblem == ContestProblems.id)) \
+        .join(Session, on=(Submission.user == Session.user))\
```
hmm. this isn't a direct foreign key. perhaps we should change the model? `submission.session.user`?

rishabhKalakoti replied:
Adding a session as foreign key to submission doesn't look right to me.
Eg.
User logged in. Made a submission with session Id = 12345. Logged out. (That session entry would be deleted from the database on logging out).
So, when we search for that session while calculating stats, it won't be available anymore.

theSage21 commented:
Oh right. You're deleting sessions instead of deactivation. Forgot

I've lost touch with this codebase. I'll schedule some time tomorrow to
work on this

On Fri 24 May, 2019, 19:27 Rishabh Kalakoti, <notifications@github.com>
wrote:

> *@rishabhKalakoti* commented on this pull request.
> ------------------------------
>
> In server.py
> <https://github.com/PyJaipur/PyJudge/pull/131#discussion_r287371122>:
>
> > @@ -146,6 +143,20 @@ def dashboard():
>      contests = Contest.select().order_by(Contest.start_time)
>      return bottle.template("dashboard.html", contests=contests)
>
> +@app.get("/stats")
> +@login_required
> +def statistics():
> +    sub_history_temp = Submission.select(Contest.code, ContestProblems.question, Submission.time, Submission.is_correct)\
> +        .where(Session.token == bottle.request.get_cookie("s_id"))\
> +        .join(ContestProblems, on=(Submission.contestProblem == ContestProblems.id)) \
> +        .join(Session, on=(Submission.user == Session.user))\
>
> Adding a session as foreign key to submission doesn't look right to me.
> Eg.
> User logged in. Made a submission with session Id = 12345. Logged out.
> (That session entry would be deleted from the database on logging out).
> So, when we search for that session while calculating stats, it won't be
> available anymore.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/PyJaipur/PyJudge/pull/131?email_source=notifications&email_token=AB2WHULWPCTVB6WZZ632R6DPW7X35A5CNFSM4HPGGX2KYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOBZUOMDQ#discussion_r287371122>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB2WHUKNQICI5WZM2DKZFVTPW7X35ANCNFSM4HPGGX2A>
> .
>

Raul9595 commented:
Ok so I should be getting the user ID from bottle and using that to reference the records right? I don't know exactly how. I'll figure it out tomorrow.

rishabhKalakoti commented:
Maybe use something like:
```
sub_history_temp = Submission.select(
    ...
)
.where(
    Submission.user == Session.get(Session.token == bottle.request.get_cookie("s_id")).user
)
...
```

Raul9595 commented:
I have deployed the code with the changes that Rishabh suggested.

rishabhKalakoti commented on server.py:
```diff
@@ -146,6 +143,20 @@ def dashboard():
+        .where(Submission.user == Session.get(Session.token == bottle.request.get_cookie("s_id")).user)\
+        .join(ContestProblems, on=(Submission.contestProblem == ContestProblems.id)) \
+        .join(Session, on=(Submission.user == Session.user))\
+        .switch() \
```
@Raul9595 You don't need to join the Session table anymore, the where statement should be enough.

rishabhKalakoti replied:
We can even avoid joining `contest`.
Refer: https://github.com/PyJaipur/PyJudge/pull/126#discussion_r285917308
Using `backref` and with the foreign key, you can directly refer to `contest` from `contestProblems`.
But wait for @theSage21 approval for this one, he might have better suggestions.

theSage21 replied:
you could. I think this query might be more clear. I tend to use the backref feature sparingly since it introduces db requests without you realizing it at first glance.

Raul9595 replied:
I have not used backref against the recommendation of @theSage21. I have removed the Session table join though.

theSage21 commented on server.py:
```diff
@@ -146,6 +143,20 @@ def dashboard():
+@login_required
+def statistics():
+    sub_history_temp = Submission.select(Contest.code, ContestProblems.question, Submission.time, Submission.is_correct)\
+        .where(Submission.user == Session.get(Session.token == bottle.request.get_cookie("s_id")).user)\
```
What I usually do is in the login_required function attach a user object called `me` to the bottle request. Then "how" I obtain the `me` is defined in one place and others can use it like `bottle.request.me`

Raul9595 replied:
I'm sorry @theSage21 but then I don't exactly understand how to do this.

theSage21 replied:
so perhaps this is how I would write the login_required function:

```python
def login_required(api):
    def wrapper(*a, **kw):
        try:
            session = Session.get(Session.token == bottle.request.get_cookie('s_id'))
        except Session.DoesNotExist:
            return abort(403, 'please log in')
        bottle.request.session = session
        return api(*a, **kw)
    return wrapper
```

then in the code:

```python

@app.get('/some/api')
@login_required
def my_api():
    SomeTable.select().join(Something).where(Session.id == bottle.request.session, User.id == bottle.request.session.user)
``

theSage21 replied:
this ensures that the logic you want to implement to get the cookie is in one place. For example when writing APIs it is common to have tokens in the `Authorization` header OR the cookie. Then this kind of code becomes useful.

Raul9595 replied:
Thanks a lot.

theSage21 commented on server.py:
```diff
@@ -146,6 +143,20 @@ def dashboard():
+        .switch() \
+        .join(Contest, on=(ContestProblems.contest == Contest.id)) \
+        .order_by(Submission.time.desc())
+    sub_history = list(sub_history_temp.tuples())
```
can we get rid of the intermediate temp var? Just put it directly in the expected variable?

theSage21 replied:
also, since we are using this later on in both template and the code, it might be a good idea to do `dicts()` instead of `tuples()`. Makes the later code clearer:

```python
sub_stats_correct = len([sub_history for sub in sub_history if sub['is_correct']])
```

theSage21 commented on server.py:
```diff
@@ -146,6 +143,20 @@ def dashboard():
+        .order_by(Submission.time.desc())
+    sub_history = list(sub_history_temp.tuples())
+    sub_stats_total = len(sub_history)
+    sub_stats_correct = len([sub_history for sub in sub_history if sub[3]==True])
```
```suggestion
    sub_stats_correct = len([sub_history for sub in sub_history if sub[3]])
```

Just use the boolean. no need to compare to another boolean

theSage21 commented on views/stats.html:
```diff
@@ -0,0 +1,84 @@
+				</tr>
+				% for sub_history in sub_history :
+				<tr>
+					<td style="padding-right:10rem">{{sub_history[0]}}</td>
```
```suggestion
					<td style="padding-right:10rem">{{sub_history.code}}</td>
```

theSage21 replied:
these kinds of changes make it more readable. Will have to make it a dictionary instead though

Raul9595 commented:
Updated the code as per the comments given above, @rishabhKalakoti and @theSage21

Raul9595 commented:
I have formatted the code using Black
